### ISSUE TITLE
Unexpected module nesting behavior when renaming a symbol

### ISSUE THREAD

**Issue by AlexEvanczuk** [sorbet/sorbet#6336]
Hi – I found this bug when exploring the ability to automatically rename space a constant with the sorbet LSP. I wasn't sure if renamespacing is a supported feature of "rename symbol" at all or if its primarily meant to be used for just changing the name of a constant without adding or removing namespaces. More thoughts on this at the bottom.

#### Input

[→ View on sorbet.run](https://sorbet.run/#%23%20typed%3A%20strict%0A%0Amodule%20Foo%0Aend%0A%0AFoo)

```ruby
# typed: strict

module Foo
end

Foo
```

#### Observed output
When clicking "rename symbol" and replacing `Foo` with `Foo::Bar`, the output is:
```
module Foo::Bar
end

Foo::Bar
```

#### Expected behavior

I'd expect the behavior to be this:
```
module Foo
  module Bar
  end
end

Foo::Bar
```

- - -

I mention above that I'm not sure if this is actually the "supported use" of rename symbol. If so, it might be really useful to have a separate feature "renamespace symbol" that can add or remove modules. This would be incredibly useful to mass reorganize code to conform to zeitwerk conventions and to also reorganize it from the perspective of domain modularity (e.g. to move it into well-factored packages where all the code lives under a specific namespace).

Here are some expected behaviors and features that I was thinking it might be able to do:
- Be able to correctly nest new/removed modules on new lines.
- Be able to remove namespaces (I don't think the current `rename symbol` can do this)
- This is a bit of a stretch, but it would be interesting if it could change file paths to match certain conventions when renaming, e.g. to match zeitwerk conventions. This behavior probably belongs in a wrapper around the LSP though I imagine.

**Comment by jez** [comment]
Do I understand you correctly:

If there is a file named `foo.rb` and it has the following code:

```ruby
module Foo
end
```

Zeitwerk can autoload `Foo` from `foo.rb`

But if we rename `Foo` to `Foo::Bar`, this is what Sorbet writes:

```ruby
module Foo::Bar
end
```

meaning that Zeitwerk cannot autoload `Foo` or `Foo::Bar`, because `Foo` is not declared anymore. Sorbet would either need to write out

```ruby
module Foo
  module Bar
  end
end
```

or to do what it currently does and also rename the file to `foo/bar.rb`.

Is that what you're saying?

Some follow up questions:

- I'm not sure how often people want the "edit current file" vs "move to new file" behavior.
- The behavior Sorbet currently does is still valid in codebases that don't use zeitwork (which might be like small gem codebases that manually `require` files explicitly). We'll have to figure out how to make any new functionality here optional.

Thanks for the example.

**Comment by froydnj** [comment]
Can we keep the current behavior but add a "split namespaced symbol" refactoring?  I don't know what the right name is here, but given:

```ruby
module Foo::Bar
...
end
```

the refactoring executed on `Foo::Bar` results in:

```ruby
module Foo
  module Bar
    ...
  end
end
```

It's maybe not great that you have to find this separate refactoring, but at least it would avoid injecting "rename symbol" with Zeitwerk-specific handling, and perhaps the namespace splitting would be useful on its own?

**Comment by AlexEvanczuk** [comment]
@jez 
> Is that what you're saying?

Not exactly. There are two separate concerns: (1) avoiding `NameError` and (2) making sure zeitwerk works.
For (1), neither the current behavior or the new behavior creates a `NameError` except in one case:
```
# This has no NameError!
module Foo
  class Bar
  end
end

module Foo
  class Baz
    puts "Module.nesting (2): #{Module.nesting.inspect}"
    Bar
  end
end

# This throws `uninitialized constant Foo::Baz::Bar` because we're not actually "in" the `Foo` namespace.
module Foo
  class Bar
  end
end

module Foo::Baz
  puts "Module.nesting (2): #{Module.nesting.inspect}"
  Bar
end
```

In this way, the current behavior changes the way Ruby constant lookup works since it changes `Module.nesting`.

For (2), Zeitwerk relies solely on the location of the file mapping to the namespace. From my understanding, zeitwerk actually doesn't care if the nesting is all on one line or is broken up over several lines -- it only cares that the file name maps to the constant.

I'm not sure if changing the file path should even be a feature here at all, especially since you can use sorbet without zeitwerk. If we think it should be, it probably should be only an option (not sure what the ideal API should be). My immediate take here is it's probably zeitwerk that would own the API to determine what the file path should be for code that defines a constant, and a separate tool could combine sorbet's rename symbol LSP function and the zeitwerk API to rename space a constant and move it all at once (that's the tool we want to build!).

@froydnj 
One option would be if "rename symbol" actually blows up if someone tries to add or remove a namespace, and instead we add a `change constant nesting` function to add or remove namespaces, and by default that always splits nesting across new lines. Therefore there are two separate operations: changing a portion of the namespace and adding/removing a namespace.

On the other hand, I'm not sure if folks intuitively think of renaming a symbol and changing its namespace as separate operations. Today, when you click "rename symbol" you only rename a portion of the namespace, e.g. `Foo::Bar::Baz`, you only modify one at a time, and you can't remove one of them. If instead, rename symbol operated on the fully qualified constant, then you could rename and rename space as a single cohesive/transactional operation that properly split namespaces across lines.

Let me know what you think!
